chunk/schema.go

```diff
@@ -573,6 +588,71 @@ func (v5Entries) GetReadMetricLabelValueEntries(_, _ uint32, tableName, hashKey
 	}, nil
 }
 
+// v6Entries fixes issues with v5 time encoding being wrong (see #337), and
+// moves label value out of range key (see #199).
+type v6Entries struct{}
+
+func (v6Entries) GetWriteEntries(_, through uint32, tableName, hashKey string, labels model.Metric, chunkID string) ([]IndexEntry, error) {
+	chunkIDBytes := []byte(chunkID)
+	encodedThroughBytes := encodeTime(through)
+
+	entries := []IndexEntry{
+		{
+			TableName:  tableName,
+			HashValue:  hashKey,
+			RangeValue: buildRangeKey(encodedThroughBytes, nil, chunkIDBytes, rangeKeyV3),
+		},
+	}
+
+	for key, value := range labels {
+		if key == model.MetricNameLabel {
+			continue
+		}
+		entries = append(entries, IndexEntry{
+			TableName:  tableName,
+			HashValue:  hashKey + ":" + string(key),
+			RangeValue: buildRangeKey(encodedThroughBytes, nil, chunkIDBytes, rangeKeyV5),
+			Value:      []byte(value),
+		})
+	}
+
+	return entries, nil
+}
+
+func (v6Entries) GetReadMetricEntries(_, through uint32, tableName, hashKey string) ([]IndexEntry, error) {
+	encodedThroughBytes := encodeTime(through)
+	return []IndexEntry{
+		{
+			TableName:  tableName,
+			HashValue:  hashKey,
+			RangeValue: buildRangeKey(encodedThroughBytes),
+		},
+	}, nil
+}
+
+func (v6Entries) GetReadMetricLabelEntries(_, through uint32, tableName, hashKey string, labelName model.LabelName) ([]IndexEntry, error) {
+	encodedThroughBytes := encodeTime(through)
+	return []IndexEntry{
+		{
+			TableName:  tableName,
+			HashValue:  hashKey + ":" + string(labelName),
+			RangeValue: buildRangeKey(encodedThroughBytes),
+		},
+	}, nil
+}
+
+func (v6Entries) GetReadMetricLabelValueEntries(_, through uint32, tableName, hashKey string, labelName model.LabelName, labelValue model.LabelValue) ([]IndexEntry, error) {
+	encodedThroughBytes := encodeTime(through)
+	return []IndexEntry{
+		{
+			TableName:  tableName,
+			HashValue:  hashKey + ":" + string(labelName),
+			RangeValue: buildRangeKey(encodedThroughBytes),
+			Value:      []byte(labelValue),
```

Comment:
Is it necessary to set the `Value` field for reads here? As far as I could find out, it's only ever read for writes (in https://sourcegraph.com/github.com/weaveworks/cortex@2278819b93fc3f63dd48b677f5325b0c514f8742/-/blob/chunk/schema.go#L57:2-57:7).